include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_utils.h

```diff
@@ -731,8 +759,18 @@ class __future : private std::tuple<_Args...>
         return __storage.__wait_and_get_value(__my_event);
     }
 
+    constexpr std::pair<std::size_t, std::size_t>
+    __wait_and_get_value(const std::shared_ptr<__result_and_scratch_storage_base>& __p_storage)
```

Comment:
So this is basically written only for the specific case where you want to return a pair of `size_t`, as well as all the helpers beneath it. 
Is there no better, more general way, to do this in `__future` / `__return_and_scratch_storage`, and allow the algorithm specifics to package it accordingly?
Could we not use / extend the old interface for `__return_and_scratch_storage` which allowed an index argument to `__get_value()` and add some function to future to request multiple elements from future, packed into a tuple of the requested length?

something like 

`std::tuple<std::size_t, std::size_t> ret = __future.get_multiple<2>();`

Reply:
I know that we need to refactor `future` and `result_and_scratch _storage`, as well as separating keepalives from real futures, etc, but this feels too tailored to the very specific needs of the algorithm.

Reply:
std::tuple<std::size_t, std::size_t> is more generic, agree

Reply:
I don't like the idea of merging as it is, but if we can commit to at least improving this part to a general solution for 2022.9.0, I wont insist on fixing it here, given that this is a late review, I think it works, and a more general solution is non-trivial.  

I've added some comments to #2003, and tagged it with 2022.9.0.  

I'd still suggest to not remove the existing infrastructure from `__return_and_scratch_storage` for an index on return values within this PR, unless there is some reason why it is explicitly required.

Reply:
If this plan sounds good to you, let me know, and we can resolve this discussion.

Reply:
> I'd still suggest to not remove the existing infrastructure from `__return_and_scratch_storage` for an index on return values within this PR

What exact code lines do you mean?

Reply:
> If this plan sounds good to you, let me know, and we can resolve this discussion

This plan sounds good.